cmd/kubeops/main_test.go

```diff
@@ -77,9 +78,27 @@ func TestParseClusterConfig(t *testing.T) {
 						CertificateAuthorityData: "ca-cert-data-additional\n",
 					},
 				},
+				PinnipedProxyURL: "http://kubeapps-internal-pinniped-proxy.kubeapps:3333",
+			},
+		},
+		{
+			name:       "parses a cluster with pinniped token exchange",
+			configJSON: `[{"name": "cluster-2", "apiServiceURL": "https://example.com", "certificateAuthorityData": "Y2EtY2VydC1kYXRhCg==", "serviceToken": "abcd", "pinnipedConfig": {"exchangeCredentials": true}}]`,
+			expectedConfig: kube.ClustersConfig{
+				Clusters: map[string]kube.ClusterConfig{
+					"cluster-2": {
+						Name:                     "cluster-2",
+						APIServiceURL:            "https://example.com",
+						CertificateAuthorityData: "ca-cert-data\n",
+						ServiceToken:             "abcd",
+						PinnipedConfig: kube.PinnipedConciergeConfig{
+							ExchangeCredentials: true,
+						},
+					},
+				},
+				PinnipedProxyURL: "http://kubeapps-internal-pinniped-proxy.kubeapps:3333",
```

Comment:
would it be better to store the URL with the rest of configuration? That way everything is together. I found it a bit confusing to have a defined proxy URL if later on it's disabled in the `PinnipedConfig`

Reply:
Yeah, the example here in the test doesn't really show why it's required for a multi-cluster setup. In such a setup, each cluster can have `pinniped-concierge` (not `-proxy`) installed, potentially in a different namespace with a different type of authenticator - or not... one of the clusters may also be configured with the OIDC options and not need pinniped. That's why we need per cluster `PinnipedConfig`. But there will only ever be one `PinnipedProxyURL` for a Kubeapps installation, no matter how many clusters there are, since it is proxying requests to all the clusters which are configured with pinniped (using each cluster's pinniped-concierge to exchange credentials).

So it didn't make sense to have the `PinnipedProxyURL` within the per-cluster `PinnipedConfig`, as that would enable invalid configurations which we cannot handle. I would have preferred for this to be `values.pinnipedProxy.URL` (or `.Endpoint`) or similar, but I also needed to be able to feed the value down through the kubeops CLI to the `pkg/kube` handlers, which only receive the cluster config. So rather than adding yet another arg, I placed it here in the config.

Perhaps the ideal here would be to make it part of `values.pinnipedProxy.URL` and use that to populate the `ClustersConfig.PinnipedProxyURL` (and making the field omitted from the json, so it isn't ever passed in as part of the clusters config). Let me know what you think, happy to update in a further PR.

Reply:
> So it didn't make sense to have the PinnipedProxyURL within the per-cluster PinnipedConfig, as that would enable invalid configurations which we cannot handle.

I get the point that having different  URLs per cluster is not a valid configuration but I still think that having all the Pinniped configuration in the same place (from this service POV, it would be the `PinnipedConfig`).

> Perhaps the ideal here would be to make it part of values.pinnipedProxy.URL and use that to populate the ClustersConfig.PinnipedProxyURL (and making the field omitted from the json, so it isn't ever passed in as part of the clusters config). Let me know what you think, happy to update in a further PR.

I've lost track of the places in which you configure this but yes, the chart values is the right place to enforce a single URL (and any other info) for later generate a `Config` per cluster if needed.